lib/android/app/src/main/java/com/reactnativenavigation/anim/TopbarCollapsingBehavior.java

```diff
@@ -0,0 +1,81 @@
+package com.reactnativenavigation.anim;
+
+
+import android.view.View;
+
+import com.facebook.react.uimanager.events.EventDispatcher;
+import com.reactnativenavigation.interfaces.ScrollEventListener;
+import com.reactnativenavigation.utils.UiThread;
+import com.reactnativenavigation.views.TopBar;
+
+public class TopbarCollapsingBehavior {
+    private TopBar topBar;
+
+    private EventDispatcher eventDispatcher;
+    private ScrollEventListener scrollEventListener;
+    private boolean dragStarted;
+    private TopBarAnimator animator;
+
+    public TopbarCollapsingBehavior(EventDispatcher eventDispatcher, TopBar topBar) {
+        this.eventDispatcher = eventDispatcher;
+        this.topBar = topBar;
+        this.animator = new TopBarAnimator();
+    }
+
+    public void enableCollapsing() {
+        scrollEventListener = (new ScrollEventListener(new ScrollEventListener.OnVerticalScrollListener() {
+            @Override
+            public void onVerticalScroll(int scrollY, int oldScrollY) {
+                if (scrollY < 0) return;
+                if (!dragStarted) return;
+
+                int measuredHeight = topBar.getMeasuredHeight();
+                int diff = scrollY - oldScrollY;
```

Comment:
Again, we can add a final modifier to `diff` as well by extracting this calculation to a dedicated method - `calculateDeltaScroll(scrollY, oldScrollY)`

Having two final variables and a helper function in a 20 line function are a good indication that this function should be extracted into a dedicated class.